src/modules/ekf2/EKF/aid_sources/optical_flow/optical_flow_control.cpp

```diff
@@ -222,13 +223,59 @@ void Ekf::controlOpticalFlowFusion(const imuSample &imu_delayed)
 	}
 }
 
-void Ekf::resetFlowFusion(const flowSample &flow_sample)
+void Ekf::resetFlowFusion(const flowSample &flow_sample, VectorState &H)
 {
+	// dont force hard reset if innovation is acceptable
+	if (!_aid_src_optical_flow.innovation_rejected) {
+		Vector2f v_xy_pre = _state.vel.xy();
+		fuseOptFlow(H, _control_status.flags.opt_flow_terrain);
+
+		// record velocity diff for output-predictor because it does not correspond to IMU measurements
+		Vector2f diff = _state.vel.xy() - v_xy_pre;
+		recordHorizontalVelReset(diff);
+		return;
+	}
+
 	ECL_INFO("reset velocity to flow");
 	_information_events.flags.reset_vel_to_flow = true;
 
 	const float flow_vel_var = sq(predictFlowRange()) * calcOptFlowMeasVar(flow_sample);
-	resetHorizontalVelocityTo(getFilteredFlowVelNE(), flow_vel_var);
+	const float epsilon = 1e-3f;
+	Vector2f innov_var_vector;
+	sym::ComputeFlowXyInnovVarAndHx(_state.vector(), P, flow_vel_var, epsilon, &innov_var_vector, &H);
+
+	const auto state_vector = _state.vector();
+	VectorState Kfusion;
+	Vector2f vel_desired = getFilteredFlowVelNE();
+	const Vector3f flow_gyro_corrected = _flow_sample_delayed.gyro_rate - _flow_gyro_bias;
+	Vector2f innov = predictFlow(flow_gyro_corrected) - vel_desired;
+
+	// there is a cross-correlation built up between the attitude/gyro-bias states and the position
+	// uncertainties which causes unstable behavior when not reset
+	P.slice<State::pos.dof, State::quat_nominal.dof>(State::pos.idx, State::quat_nominal.idx) = 0.f;
+	P.slice<State::quat_nominal.dof, State::pos.dof>(State::quat_nominal.idx, State::pos.idx) = 0.f;
+	P.slice<State::pos.dof, State::gyro_bias.dof>(State::pos.idx, State::gyro_bias.idx) = 0.f;
+	P.slice<State::gyro_bias.dof, State::pos.dof>(State::gyro_bias.idx, State::pos.idx) = 0.f;
+
+	for (int index = 0; index < 2; index++) {
+		P.uncorrelateCovariance<1>(State::vel.idx + index);
+
+		if (index == 1) {
+			const float R_LOS = flow_vel_var;
+			sym::ComputeFlowYInnovVarAndH(state_vector, P, R_LOS, epsilon, &innov_var_vector(1), &H);
```

Comment:
That doesn't really work out, you're mixing flow axes (body frame) with NED axes.